File: 백준/Silver/2346. 풍선 터뜨리기/풍선 터뜨리기.py
from collections import deque
import sys
input = sys.stdin.readline

N = int(input())
nums = deque(enumerate(map(int, input().split()), start=1))

i = 0
result = []
while nums:
    idx, move = nums.popleft()
    result.append(idx)

    if not nums:
        break
    if move > 0:
        move -= 1
    nums.rotate(-move)
print(' '.join(map(str, result)))

File: 백준/Silver/2346. 풍선 터뜨리기/풍선 터뜨리기.java
import java.io.*;
import java.util.*;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        int N = Integer.parseInt(br.readLine());
        StringTokenizer st = new StringTokenizer(br.readLine());
        LinkedList<int[]> nums = new LinkedList<>();

        for (int i = 1; i <= N; i++) {
            nums.add(new int[]{i, Integer.parseInt(st.nextToken())});
        }

        List<Integer> result = new ArrayList<>();

        while (!nums.isEmpty()) {
            int[] current = nums.pollFirst();

            int idx = current[0];
            int move = current[1];
            result.add(idx);

            if (nums.isEmpty()) {
                break;
            }

            if (move > 0) {
                move--;
            }

            Collections.rotate(nums, -move);
        }

        StringBuilder sb = new StringBuilder();
        for (int n : result) {
            sb.append(n).append(" ");
        }
        System.out.println(sb.toString().trim());
    }
}